### ISSUE TITLE
Precondition, postconditions and invariants with treat.js

### ISSUE THREAD

**Issue by brucou** [keil/TreatJS#5]
Hi there,

i am looking for a flexible and usable library for specifying run-time contracts in my js code. There are a number of options, most of which involve altering the source code.

I have a few questions regarding treat:
- the contracts you build seem to be constructed (`Base`, `SFunction` for instance) from predicates and then 
there are some combinators which allow to build more complex contracts from simple one. Are contracts monads? This might open a way to build combinators with `map`, `flatMap` and `return`.
- all the examples I see seem to be checking inputs of functions. What about outputs? What about invariants? Is there a way to do this with treat.js?
- you insist on the sandboxing of the library. It would be great if you could explain in layman terms what that is, and the benefits associated to it.
- this url `http://proglang.informatik.uni-freiburg.de/treatjs/treatjs.config.html` results in `You don't have permission to access /treatjs/treatjs.config.html on this server.`. Is that the expected behaviour?

Quite interesting in any case to see some academia behind this,

**Comment by keil** [comment]
Hi brucou

First, I recommend that you look to TreatJS 2.0, instead of TreatJS 1.*.

Yes, most systems implement “contracts” in form of an offline code transformation. TreatJS uses JS proxies to enforce contracts at run time. However, you must be aware of the fact that introducing contracts might change the run time behaviour of your program, if you do not use an JS engine that provides transparent proxies. You might habe a look to this paper to understand why: http://drops.dagstuhl.de/opus/volltexte/2015/5229/

> the contracts you build seem to be constructed (Base, SFunction for instance) from predicates and then there are some combinators which allow to build more complex contracts from simple one. 

Base contracts (aka flat contracts) and constructors are the only fundamental building blocks for contracts. Base contracts consist of predicates, constructors of constructor functions. All other contracts  (including function contracts) are composed of other contracts. In the simplest way of base contracts. And, the SFunction contract is only a short form of a normal function contract.

> Are contracts monads? This might open a way to build combinators with map, flatMap and return.

In some way, yes. 

> all the examples I see seem to be checking inputs of functions. What about outputs? What about invariants? Is there a way to do this with treat.js?

Thats not correct. Each function contract consist of one contract for the domain, and one contract for the range of a function. So, they always check both, input and return of a function. Invariants are technically possible, but right now not implemented in TreatJS 2.0 (there is an draft implementation in TreatJS 1.*)

> you insist on the sandboxing of the library. It would be great if you could explain in layman terms what that is, and the benefits associated to it.

In short, the sandbox ensures that the execution of predicates (which are normal JS function) did not influence the normal program execution. They were executed in some kind of restricted area and cannot touch the normal execution.

> this url http://proglang.informatik.uni-freiburg.de/treatjs/treatjs.config.html results in You don't have permission to access /treatjs/treatjs.config.html on this server.. Is that the expected behaviour?

I’m sorry for this. It seems to be a bug in our CMS.

> Quite interesting in any case to see some academia behind this,

?

- Matthias
